### PR TITLE
Fix #4119, Fix #4120 - only re escape for re search. Fix #3571 or at …

### LINKED ISSUE #3571: "regular expression is invalid: \\ at end of pattern" error still not fixed.
```
2022-09-06 18:46:23,386 - [scout.server.app](http://scout.server.app/) - ERROR: Exception on /cust003/cases [GET] [in /venv/lib/python3.8/site-packages/flask/app.py:1741]
Traceback (most recent call last):
  File "/venv/lib/python3.8/site-packages/flask/app.py", line 2525, in wsgi_app
    response = self.full_dispatch_request()
  File "/venv/lib/python3.8/site-packages/flask/app.py", line 1822, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/venv/lib/python3.8/site-packages/flask_cors/extension.py", line 165, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/venv/lib/python3.8/site-packages/flask/app.py", line 1820, in full_dispatch_request
    rv = self.dispatch_request()
  File "/venv/lib/python3.8/site-packages/flask/app.py", line 1796, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
  File "/home/worker/app/scout/server/utils.py", line 85, in decorated_function
    context = func(*args, **kwargs)
  File "/home/worker/app/scout/server/blueprints/institutes/views.py", line 76, in cases
    return controllers.cases(store, request, institute_id)
  File "/home/worker/app/scout/server/blueprints/institutes/controllers.py", line 399, in cases
    all_cases = store.cases(
  File "/home/worker/app/scout/adapter/mongo/case.py", line 346, in cases
    order = self._populate_name_query(query, name_query, owner, collaborator)
  File "/home/worker/app/scout/adapter/mongo/case.py", line 235, in _populate_name_query
    query["assignees"] = {"$in": [user["email"] for user in users]}
  File "/home/worker/app/scout/adapter/mongo/case.py", line 235, in <listcomp>
    query["assignees"] = {"$in": [user["email"] for user in users]}
  File "/venv/lib/python3.8/site-packages/pymongo/cursor.py", line 1238, in next
    if len(self.__data) or self._refresh():
  File "/venv/lib/python3.8/site-packages/pymongo/cursor.py", line 1155, in _refresh
    self.__send_message(q)
  File "/venv/lib/python3.8/site-packages/pymongo/cursor.py", line 1044, in __send_message
    response = client._run_operation(
  File "/venv/lib/python3.8/site-packages/pymongo/mongo_client.py", line 1424, in _run_operation
    return self._retryable_read(
  File "/venv/lib/python3.8/site-packages/pymongo/mongo_client.py", line 1525, in _retryable_read
    return func(session, server, sock_info, secondary_ok)
  File "/venv/lib/python3.8/site-packages/pymongo/mongo_client.py", line 1420, in _cmd
    return server.run_operation(
  File "/venv/lib/python3.8/site-packages/pymongo/server.py", line 130, in run_operation
    _check_command_response(first, sock_info.max_wire_version)
  File "/venv/lib/python3.8/site-packages/pymongo/helpers.py", line 167, in _check_command_response
    raise OperationFailure(errmsg, code, response, max_wire_version)
pymongo.errors.OperationFailure: Regular expression is invalid: \ at end of pattern, full error: {'operationTime': Timestamp(1662482778, 1), 'ok': 0.0, 'errmsg': 'Regular expression is invalid: \\ at end of pattern', 'code': 51091, 'codeName': 'Location51091', '$clusterTime': {'clusterTime': Timestamp(1662482778, 1), 'signature': {'hash': b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00', 'keyId': 0}}}
```

Perhaps we should introduce a debug message to try to understand what they are typing, to reproduce the error?

### LINKED ISSUE #4119: Patient cohorts with spaces cannot be searched
Patient cohort tags with spaces won't match their cases on the caseS overview.
Screenshot 2023-10-12 at 08 11 23

Tentatively the search string is re.escaped, but not re-searched in the db query, or converted back to plain text.

### LINKED ISSUE #4120: HPO term searches order dependent
Here's another casualty to the `re.escape()` that we can fix. When searching for multiple HPO terms, only the first is guaranteed to actually be searched.

Reproduce by setting a phenotype on a case:
Screenshot 2023-10-12 at 09 34 37

Search for two phenotypes with the matching one first
Screenshot 2023-10-12 at 09 34 29

Then second:
Screenshot 2023-10-12 at 09 34 12

I believe this is related to #4119.

### PR DESCRIPTION
**Pull request by dnil**
…least harden against it. It may not be deriving from direct input into the caseS search box.

This PR adds a functionality or fixes a bug.

Testing details here: 
Fix #4119 
Fix #4120 
 Briefly, try HPO terms in swapped order, and cohort tags with spaces.

Before:
Screenshot 2023-10-12 at 11 41 50

Screenshot 2023-10-12 at 09 34 12

Screenshot 2023-10-12 at 11 47 18

Screenshot 2023-10-12 at 11 46 07

After:
Screenshot 2023-10-12 at 11 58 29

Screenshot 2023-10-12 at 11 57 45

Screenshot 2023-10-12 at 11 55 09

For #3571, note that `\` is now disallowed in search field, in contrast to current main.

Before:
Screenshot 2023-10-12 at 11 41 26

After:
Screenshot 2023-10-12 at 11 27 14

I suspect some linter may have issues with the handling of direct user input. As per my reading, we are safe (the input appears as strings in different subqueries and should not have the opportunity to break out of them without causing error), but let's see. Some of the query categories could be made safer by only allowing existing values, but then we need to track that..

<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. Fist book your testing time using the Pax software available at [https://pax.scilifelab.se/](https://pax.scilifelab.se). The resource you are going to call dibs on is `scout-stage` and the server is `cg-vm1`.
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
1. After testing is done, repeat procedure at [https://pax.scilifelab.se/](https://pax.scilifelab.se), which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

<details>
<summary>Testing on hasta server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. `ssh <USER.NAME>@hasta.scilifelab.se`
1. Book your testing time using the Pax software. `us; paxa -u <user> -s hasta -r scout-stage`. You can also use the WSGI Pax app available at [https://pax.scilifelab.se/](https://pax.scilifelab.se).
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `conda activate S_scout; pip freeze | grep scout-browser`
1. Deploy the branch to test: `bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_scout -t scout -b <this_branch>`
1. Make sure the branch is deployed: `us; scout --version`
1. After testing is done, repeat the `paxa` procedure, which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

**How to test**:
1. how to test it, possibly with real cases/data

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [x] code approved by CR
- [x] tests executed by DN

### CODE REVIEW

File: scout/adapter/mongo/case.py [hunk 1]
```diff
@@ -151,10 +149,11 @@ def _set_genes_of_interest_query(
     def _set_case_name_query(self, query: Dict[str, Any], query_term: str):
         """Set case query to reg exp search in case and individual display names for parts of the name query."""
 
+        escaped_query_term = re.escape(query_term)
```
Comment (dnil) [inline scout/adapter/mongo/case.py:152]: RE-escaping will work well here, since the term is used for RE-searching.

Comment (northwestwitch) [inline scout/adapter/mongo/case.py:152]: 💯

File: scout/adapter/mongo/case.py [hunk 2]
```diff
@@ -222,6 +221,8 @@ def _populate_name_query(self, query, name_query, owner=None, collaborator=None)
         query_field = name_query.split(":")[0]  # example:status
         query_term = name_query[name_query.index(":") + 1 :].strip()
 
+        LOG.info("Name query: %s Field: %s term: %s", name_query, query_field, query_term)
```
Comment (northwestwitch) [inline scout/adapter/mongo/case.py:224]:
```suggestion
```

File: scout/server/blueprints/institutes/templates/overview/cases.html [hunk 3]
```diff
@@ -30,7 +30,7 @@
       </div>
       <div class="col-3">
           <label class="control-label" for="search_term">Search cases</label>
-          <input class="form-control" id="search_term" name="search_term" value="{{form.search_term.data}}" placeholder="example:18201" type="text" pattern="[^<>?!=\\/]*" title="Characters ^<>?!=\/ are not allowed">
+          <input class="form-control" id="search_term" name="search_term" value="{{form.search_term.data}}" placeholder="example:18201" type="text" pattern="[^\\\<\>\?\!\=\/]*" title="Characters \<>?!=/ are not allowed">
```
Comment (northwestwitch) [inline scout/server/blueprints/institutes/templates/overview/cases.html:33]: Woa!

**Review by northwestwitch — APPROVED** [review]
Great fix! 🥇 It's adding a lot of value to cases search!

File: scout/adapter/mongo/case.py (hunk 2 above)
Comment (dnil) [reply scout/adapter/mongo/case.py]: Hehe, I put it back again when reading the comments on #3571, but sure.

**Comment by northwestwitch** [comment]
Since you are at it you could perhaps address the sonarcloud's code smells? They look easy enough to fix

**Pull request merged**